### DISCUSSION TITLE
[Proposal] Abstract interfaces

### DISCUSSION THREAD

**Discussion by WhitWaldo** [dotnet/csharplang#7297] [Language Ideas] [answered]
One can mark a class as abstract to force developers to implement it in another class to utilize it and ensuring that there are no instances of it initialized anywhere.

I would like to propose that the same be able to be done with interfaces. The idea is that an interface marked as abstract can only be extended by another interface but cannot itself be directly implemented. Why? To prevent developers from using base interfaces that are intended to denote core functionality and force either that they implement their own interface or utilize any of the provided extended interfaces.

As an example:

```cs
public abstract interface IBaseInterface
{
}

//Legal
public interface IExtendedInterface : IBaseInterface
{
}

//Legal
public class ExtendedBase : IExtendedInterface
{
}

//Illegal
public class Base : IBaseInterface
{
}

//Illegal
public class Base : IExtendedInterface, IBaseInterface
{
}

```

Thank you for the consideration!

**Comment by CyrusNajmabadi** [comment, marked as answer]
Hi @WhitWaldo!  One way to accomplish this today would be to define an attribute for this purpose, and then an analyzer which errors if someone runs afoul of these rules you want.  You could do that today, and it would work on all versions of the language, and going back to any Roslyn compiler which supports analyzers (which goes back a long ways) :-)

**Reply by WhitWaldo** [reply]
Oh, that's a nifty idea - thank you for the suggestion!

**Comment by DavidArno** [comment]
> To prevent developers from using base interfaces that are intended to denote core functionality and force either that they implement their own interface or utilize any of the provided extended interfaces.

Why would you want to do that? What if that base interface already contains all of the contract that I'm interested in? I'd be forced to do:
```cs
public abstract interface IBaseInterface
{
    ... // defines everything I'm interested in
}

public interface IMyOughtToBeUnnecessaryInterface : IBaseInterface {}

public MyClass : IMyOughtToBeUnnecessaryInterface 
{
    ....
}
```
I can't think of a good use case for forcing that extra step on people.

**Reply by CyrusNajmabadi** [reply]
It's common to have a core interface (sometimes even empty), with derived interface that define, effectively, a discriminated union of domain values. 

The core interface exists to help ensure that not any object can be passed along, but specifically only domain objects. But in this modeling, it would not make sense to allow an impl of the base interface. 

It's the same reason why you can't instantiate an abstract base class, even if there are no abstract members in it itself. It's very abstractness forces only instantiating subtypes. This is that same concept, just lifted to interfaces. 

Currently the language itself has no enforcement of this (but it would make sense for certain domains). But, as this is a request for domain restrictions, attributes/analyzers is a sensible choice, and it seems like that answered the question well.

**Reply by WhitWaldo** [reply]
This is precisely what I had in mind in that for the purpose of my own library, I intend only for interfaces that extend this core interface to actually be used anywhere and don't have any logic that allows only this "base" interface, so it wouldn't make sense for the developer to implement the base interface directly as nothing would allow it. Rather, because I intend only for interfaces that extend on this base to be used, it made sense to me that it'd be similar to that of base classes, that by marking them abstract, I can provide guidance to the implementing developer about how I intend their usage of the library to proceed.

But @CyrusNajmabadi provided a great alternative - I'd forgotten about analyzers in general and the fact they can throw build errors and not just warnings fills the gap sufficiently enough in this case. Between the workaround of simply implementing an analyzer and the uphill battle that is adding a language feature, it made sense to me that this discussion should be closed rather than otherwise pursued.

Of course, if there's any realized value in re-opening this for actual language consideration, I'd be happy to do that too - I just have a viable short-term path forward that doesn't necessitate that.
